ethereum/core/src/main/java/org/hyperledger/besu/ethereum/core/contract/CodeCache.java

```diff
@@ -0,0 +1,74 @@
+/*
+ * Copyright ConsenSys AG.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on
+ * an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations under the License.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package org.hyperledger.besu.ethereum.core.contract;
+
+import org.hyperledger.besu.ethereum.core.Account;
+import org.hyperledger.besu.ethereum.vm.Code;
+
+import java.util.Optional;
+
+import com.google.common.cache.CacheBuilder;
+import com.google.common.cache.LoadingCache;
+import org.checkerframework.checker.nullness.qual.Nullable;
+
+public class CodeCache {
+
+  private final LoadingCache<WrappedAccount, Code> cache;
```

Comment:
In general this WrappedAccount scheme won't work.  I would instead suggest `CodeAndHash` where the Hash and Code object are stored as fields, the `equals` method first compares the hash and the code, and the `hashCode` method only returns the `hashCode` of the `Hash`.  Use that as the key instead of any other Account object.  There can be a static factory or constructor too that takes the `Account` object.